toolchain/semantics/semantics_node.h

```diff
@@ -124,207 +124,219 @@ struct SemanticsStringId : public IndexBase {
 
-  static auto MakeCodeBlock(ParseTree::Node parse_node,
-                            SemanticsNodeBlockId node_block) -> SemanticsNode {
-    return SemanticsNode(parse_node, SemanticsNodeKind::CodeBlock,
-                         SemanticsNodeId::Invalid, node_block.index);
-  }
-  auto GetAsCodeBlock() const -> SemanticsNodeBlockId {
-    CARBON_CHECK(kind_ == SemanticsNodeKind::CodeBlock);
-    return SemanticsNodeBlockId(arg0_);
-  }
+   private:
+    // GetImpl handles the different return types based on ArgTypes.
+    template <typename Arg0Type, typename Arg1Type, typename>
+    static auto GetImpl(SemanticsNode node) -> std::pair<Arg0Type, Arg1Type> {
+      CARBON_CHECK(node.kind() == Kind);
+      return {Arg0Type(node.arg0_), Arg1Type(node.arg1_)};
+    }
+    template <typename Arg0Type, typename>
+    static auto GetImpl(SemanticsNode node) -> Arg0Type {
+      CARBON_CHECK(node.kind() == Kind);
+      return Arg0Type(node.arg0_);
+    }
+    template <typename>
+    static auto GetImpl(SemanticsNode node) -> NoArgs {
+      CARBON_CHECK(node.kind() == Kind);
+      return NoArgs();
+    }
+  };
+
+  // Provide Get along with a Make that requires a type.
+  template <KindTemplateEnum Kind, typename... ArgTypes>
+  class Factory : public FactoryBase<Kind, ArgTypes...> {
+   public:
+    using FactoryBase<Kind, ArgTypes...>::Make;
+    using FactoryBase<Kind, ArgTypes...>::Get;
+  };
+
+  // Provides Get along with a Make that assumes a non-changing type.
+  template <KindTemplateEnum Kind, int32_t TypeIndex, typename... ArgTypes>
+  class FactoryPreTyped : public FactoryBase<Kind, ArgTypes...> {
+   public:
+    static auto Make(ParseTree::Node parse_node, ArgTypes... args) {
+      SemanticsNodeId type_id(TypeIndex);
+      return FactoryBase<Kind, ArgTypes...>::Make(parse_node, type_id, args...);
+    }
+    using FactoryBase<Kind, ArgTypes...>::Get;
+  };
 
-  static auto MakeCrossReference(SemanticsNodeId type,
-                                 SemanticsCrossReferenceIRId ir,
-                                 SemanticsNodeId node) -> SemanticsNode {
-    return SemanticsNode(ParseTree::Node::Invalid,
-                         SemanticsNodeKind::CrossReference, type, ir.index,
-                         node.index);
-  }
-  auto GetAsCrossReference() const
-      -> std::pair<SemanticsCrossReferenceIRId, SemanticsNodeId> {
-    CARBON_CHECK(kind_ == SemanticsNodeKind::CrossReference);
-    return {SemanticsCrossReferenceIRId(arg0_), SemanticsNodeId(arg1_)};
-  }
+ public:
+  // Invalid is in the SemanticsNodeKind enum, but should never be used.
+  class Invalid {
+   public:
+    static auto Get(SemanticsNode /*node*/) -> SemanticsNode::NoArgs {
+      CARBON_FATAL() << "Invalid access";
+    }
+  };
+
+  class Assign : public SemanticsNode::Factory<SemanticsNodeKind::Assign,
+                                               SemanticsNodeId /*lhs_id*/,
+                                               SemanticsNodeId /*rhs_id*/> {};
```

Comment:
Would an alias work here (and elsewhere its this simple)?
```suggestion
  using Assign = SemanticsNode::Factory<SemanticsNodeKind::Assign,
                                        SemanticsNodeId /*lhs_id*/,
                                        SemanticsNodeId /*rhs_id*/>;
```

Reply:
Yes, switched the trivial classes over.